platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/AbstractCollectionProperty.java

```diff
@@ -622,57 +617,124 @@ protected Value<? extends C> calculateOwnValue(ValueConsumer consumer) {
-            Value<Void> resultingValue = Value.present();
             Value<Void> leftValue = left.collectEntries(consumer, collector, dest);
             if (leftValue.isMissing()) {
                 return leftValue;
-            } else {
-                resultingValue = resultingValue.withSideEffect(SideEffect.fixedFrom(leftValue));
             }
             Value<Void> rightValue = right.collectEntries(consumer, collector, dest);
             if (rightValue.isMissing()) {
                 return rightValue;
-            } else {
-                resultingValue = resultingValue.withSideEffect(SideEffect.fixedFrom(rightValue));
             }
-            return resultingValue;
+
+            return Value.present()
+                .withSideEffect(SideEffect.fixedFrom(leftValue))
+                .withSideEffect(SideEffect.fixedFrom(rightValue));
         }
 
         @Override
         public void calculateExecutionTimeValue(Action<? super ExecutionTimeValue<? extends Iterable<? extends T>>> visitor) {
             left.calculateExecutionTimeValue(visitor);
             right.calculateExecutionTimeValue(visitor);
         }
+    }
+
+    /**
+     * A plus collector that either produces a composition of both of its left and right sides,
+     * or Value.present().
+     */
+    private static class AbsentIgnoringPlusCollector<T> extends AbstractPlusCollector<T> {
+
+        public AbsentIgnoringPlusCollector(Collector<T> left, Collector<T> right) {
+            super(left, right);
+        }
 
         @Override
-        public ValueProducer getProducer() {
-            return left.getProducer().plus(right.getProducer());
+        public boolean calculatePresence(ValueConsumer consumer) {
+            return true;
         }
 
         @Override
-        public String toString() {
-            return left + " + " + right;
+        public Collector<T> absentIgnoring() {
+            return this;
+        }
+
+        @Override
+        public Value<Void> collectEntries(ValueConsumer consumer, ValueCollector<T> collector, ImmutableCollection.Builder<T> dest) {
+            ImmutableList.Builder<T> candidateEntries = ImmutableList.builder();
+            Value<Void> leftValue = left.collectEntries(consumer, collector, candidateEntries);
+            if (leftValue.isMissing()) {
+                return Value.present();
+            }
+            Value<Void> rightValue = right.collectEntries(consumer, collector, candidateEntries);
+            if (rightValue.isMissing()) {
+                return Value.present();
+            }
+            dest.addAll(candidateEntries.build());
+            return Value.present()
+                .withSideEffect(SideEffect.fixedFrom(leftValue))
+                .withSideEffect(SideEffect.fixedFrom(rightValue));
+        }
+
+        @Override
+        public void calculateExecutionTimeValue(Action<? super ExecutionTimeValue<? extends Iterable<? extends T>>> visitor) {
```

Comment:
@mlopatkin Mike, this was the only change I made to the set/list supporting code after your review earlier today. I had a test case for map property (which I now added to `CollectionPropertySpec`) that would fail (because I removed `ignoreAbsent` from `CollectingProvider`).